text/0083-rawkv-cross-cluster-replication.md

```diff
@@ -0,0 +1,185 @@
+
+##### 2.2.1 Physical Part of Timestamp
+
+Physical part of HLC is acquired from TSO of [PD](https://github.com/tikv/pd).
+
+TSO is a global monotonically increasing timestamp, which help the generation of timestamp be independent to local clock of machine, and be immune to issues such as reverse between machine reboot.
+
+Physical part is refreshed by repeatedly acquiring TSO in a period of `500ms`, to keep it being closed the real world time.
+
+On startup, physical part must be initialized by a successful TSO, to ensure that HLC is larger than last running. RawKV write operations will fail if the initialization has not been completed. (TiKV also can't startup when PD is out of service).
+
+After that, HLC can tolerate fault of TSO. RawKV operations are normal, but time-related metrics such as RPO is not reasonable. Meanwhile, as TiKV-CDC being highly dependent to PD (for cluster and task management, checkpoint, etc.), the replication will not be working.
+
+##### 2.2.2 Logical Part of Timestamp
+
+Logical part is advanced on every write request.
+
+Moreover, the logical part is advanced on *Leader Transfer*. As TiKV is a distributed system, every store in the TiKV cluster has an instance of timestamp generator. As every key exists in only one region and only one peer (the leader) in the region can write, it's safe to generate timestamps locally in each store, except for leader transfer.
+
+On leader transfer, the timestamp generated by the original store could be larger than the store where the new leader is located, which will violate the causal consistency requirement. We solve this by this method: Every peer observes the raft message applied and keeps the maximum timestamp of the region (called `Region-Max-TS`). To become leader, this peer advances the timestamp generator of its store to be larger than `Region-Max-TS`. 
+
+Besides, other region changes should be carefully handled:
+
+* On region merge: The `Region-Max-TS` of the region is the max of the two original regions.
+
+* On region split: The `Region-Max-TS` of the new region is copied from the original one.
+
+#### 2.3 Resolved Timestamp
+
+**TiKV-CDC** using *Resolved Timestamp* (or `resolved-ts`) to indicate the largest transaction timestamp of the replication task, to help downstream avoid getting partial data of a transaction. Downstream streaming system can also use `resolved-ts` as [*watermark*](https://www.oreilly.com/radar/the-world-beyond-batch-streaming-102/).
+
+As no transaction support, in RawKV scenario we don't need scanning locks in *lock_cf*. But we should notify `resolved-ts` event generation that the minimum timestamp of data that on the way from *proposed* to *applied* (by [`resolver.track_lock`](https://github.com/tikv/tikv/blob/v5.0.4-20211201/components/resolved_ts/src/resolver.rs#L47-L57)).
```

Comment:
> But we should notify `resolved-ts` event generation that the minimum timestamp of data that on the way from *proposed* to *applied* (by [`resolver.track_lock`](https://github.com/tikv/tikv/blob/v5.0.4-20211201/components/resolved_ts/src/resolver.rs#L47-L57)).

I'm a little confused, does it mean we need to record the minimal timestamp of data that is proposed but not applied?

Reply:
Yes. I add a diagram and improve some words, PTAL~